[css-view-transitions-2] Implement pageswap and pagereveal events.

mattwoodrow opened the pull request:
#### 9f76fdfc2789893b2ac005d7644d1e2a10622305
<pre>
[css-view-transitions-2] Implement pageswap and pagereveal events.
<a href="https://bugs.webkit.org/show_bug.cgi?id=277524">https://bugs.webkit.org/show_bug.cgi?id=277524</a>
&lt;<a href="https://rdar.apple.com/133025306">rdar://133025306</a>&gt;

Reviewed by Tim Nguyen.

Adds the basic IDLs for the pageswap and pagereveal events, plus the C++
implementations.

Adds code to fire these events (gated on the cross document view transitions
preference), and passes the view transition parameter in a partially correct
way.

This is the basic needed for the WPT test infrastructure to work, and as a
result, many new tests pass. There are still lots of FIXME&apos;s and bugs left
before the remaining tests pass.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-layer-cascade-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-layer-cascade-external-stylesheet-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-matching-media-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-non-matching-media-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-multiple-rules-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-opt-in-change-with-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/hide-before-reveal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/mismatched-snapshot-containing-block-size-skips-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-microtask-sequence-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-no-view-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-no-view-transition-new-opt-out-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-updatecallbackdone-promise-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-with-view-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-long-delay-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-skip-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-traverse-navigation-no-bfcache.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/skip-outbound-vt-before-reveal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/at-rule-with-types-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/navigation-supersedes-types-same-rule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/navigation-supersedes-types-when-after-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/types-in-pagereveal-and-pageswap-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::reveal):
(WebCore::Document::transferViewTransitionParams):
(WebCore::Document::dispatchPageSwapEvent):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/PageRevealEvent.cpp: Copied from Source/WebCore/page/NavigationActivation.h.
(WebCore::PageRevealEvent::create):
(WebCore::PageRevealEvent::PageRevealEvent):
* Source/WebCore/dom/PageRevealEvent.h: Copied from Source/WebCore/page/NavigationActivation.h.
* Source/WebCore/dom/PageRevealEvent.idl: Added.
* Source/WebCore/dom/PageSwapEvent.cpp: Copied from Source/WebCore/page/NavigationActivation.h.
(WebCore::PageSwapEvent::create):
(WebCore::PageSwapEvent::PageSwapEvent):
* Source/WebCore/dom/PageSwapEvent.h: Copied from Source/WebCore/page/NavigationActivation.h.
* Source/WebCore/dom/PageSwapEvent.idl: Added.
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::ViewTransition):
(WebCore::ViewTransition::create):
(WebCore::ViewTransition::createInbound):
(WebCore::ViewTransition::startInbound):
(WebCore::ViewTransition::callUpdateCallback):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::takeViewTransitionParams):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::swap):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::navigationCanTriggerCrossDocumentViewTransition):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::clear):
(WebCore::FrameLoader::transitionToCommitted):
* Source/WebCore/page/NavigationActivation.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/Page.h:

Canonical link: <a href="https://commits.webkit.org/282038@main">https://commits.webkit.org/282038@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80f0b4b6225f894acead2fe30ac1a90e66dea968

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61830 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41184 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65809 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12375 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63949 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48870 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12646 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49831 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8569 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64899 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38245 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53550 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30663 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34889 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10783 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11306 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56698 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11087 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67538 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5773 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10822 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57211 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5798 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53501 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57447 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13756 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4732 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38068 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39164 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/37813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/Document.h:
```diff
@@ -1380,6 +1383,8 @@ class Document
     void queueTaskToDispatchEventOnWindow(TaskSource, Ref<Event>&&);
     void dispatchPageshowEvent(PageshowEventPersistence);
     void dispatchPagehideEvent(PageshowEventPersistence);
+    void dispatchPageSwapEvent(bool canTriggerCrossDocumentViewTransition);
```
nit: consistency with casing would be nice here

`canTriggerCrossDocumentViewTransition` could also be an enum class

mattwoodrow replied:
I think this will change soon to be the set of types that the old page wants to support.

nt1m commented on Source/WebCore/dom/PageSwapEvent.idl:
```diff
@@ -0,0 +1,34 @@
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
```
Is this the right license?

achristensen07 replied:
It is not.

nt1m commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2492,6 +2493,31 @@ ShouldOpenExternalURLsPolicy DocumentLoader::shouldOpenExternalURLsPolicyToPropa
+    if (!m_triggeringAction.navigationAPIType() || *m_triggeringAction.navigationAPIType() == NavigationNavigationType::Reload)
+        return false;
+
+    Ref<SecurityOrigin> newOrigin(SecurityOrigin::create(documentURL()));
```
```suggestion
    Ref newOrigin = SecurityOrigin::create(documentURL());
```

nt1m commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -127,14 +127,27 @@ struct OrderedNamedElementsMap {
 class ViewTransition : public RefCounted<ViewTransition>, public CanMakeWeakPtr<ViewTransition>, public ActiveDOMObject {
 public:
-    static Ref<ViewTransition> create(Document&, RefPtr<ViewTransitionUpdateCallback>&&);
+    static Ref<ViewTransition> create(Document&, std::optional<RefPtr<ViewTransitionUpdateCallback>>&&);
```
For the sake of clarity, maybe separate same-page & outbound. Having  the `std::optional<RefPtr<ViewTransitionUpdateCallback>>&&` type in the constructor doesn't help readability.

```
static Ref<ViewTransition> createSamePage(Document&, RefPtr<ViewTransitionUpdateCallback>&&);

static Ref<ViewTransition> createOutbound(Document&);

static Ref<ViewTransition> createInbound(Document&, std::unique_ptr<ViewTransitionParams>)
```

nt1m commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2492,6 +2493,31 @@ ShouldOpenExternalURLsPolicy DocumentLoader::shouldOpenExternalURLsPolicyToPropa
+// https://www.w3.org/TR/css-view-transitions-2/#navigation-can-trigger-a-cross-document-view-transition
+bool DocumentLoader::navigationCanTriggerCrossDocumentViewTransition(Document& oldDocument)
+{
+    // implementation-defined navigation experience
```
```suggestion
    // FIXME: Consider adding implementation-defined navigation experience step.
```

The current comment is a bit cryptic, I would either remove it, or reword it as a FIXME

nt1m commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2492,6 +2493,31 @@ ShouldOpenExternalURLsPolicy DocumentLoader::shouldOpenExternalURLsPolicyToPropa
+    if (!newOrigin->isSameOriginAs(oldDocument.securityOrigin()))
+        return false;
+
+    // If newDocument was created via cross-origin redirects, then return false.
```
```suggestion
    // FIXME: If newDocument was created via cross-origin redirects, then return false.
```

nt1m commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2492,6 +2493,31 @@ ShouldOpenExternalURLsPolicy DocumentLoader::shouldOpenExternalURLsPolicyToPropa
+    if (*m_triggeringAction.navigationAPIType() == NavigationNavigationType::Traverse)
+        return true;
+
+    // If isBrowserUINavigation is true, then return false.
```
```suggestion
    // FIXME: If isBrowserUINavigation is true, then return false.
```

nt1m commented on Source/WebCore/page/NavigationActivation.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "NavigationHistoryEntry.h"
```
Is this used?

mattwoodrow replied:
Yeah, unified build sadness.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8080,6 +8082,57 @@ bool Document::resolveViewTransitionRule()
     return false;
 }
 
+void Document::reveal()
```
Add a link to https://html.spec.whatwg.org/multipage/browsing-the-web.html#revealing-the-document

nt1m commented on Source/WebCore/page/Page.h:
```diff
@@ -273,6 +273,7 @@ enum class RenderingUpdateStep : uint32_t {
 #endif
     RestoreScrollPositionAndViewState = 1 << 23,
     AdjustVisibility                  = 1 << 24,
+    Reveal                            = 1 << 25,
```
Would be good to have this reflect the actual order

mattwoodrow replied:
I didn't do this one, seems like it's not the case for lots of others either. I can if you feel strongly :)

nt1m replied:
It is slightly confusing to see this towards the end here but later realize it's actually one of the first steps. I think Simon likes to keep this in sync with the actual order too (although arguably we failed to do that over the years)

nt1m commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -188,7 +205,7 @@ class ViewTransition : public RefCounted<ViewTransition>, public CanMakeWeakPtr<
     float m_initialPageZoom;
 
-    RefPtr<ViewTransitionUpdateCallback> m_updateCallback;
+    std::optional<RefPtr<ViewTransitionUpdateCallback>>  m_updateCallback;
```
Maybe describe what the 3 states are, for people newly reading this code.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8080,6 +8082,57 @@ bool Document::resolveViewTransitionRule()
+
+    PageRevealEvent::Init init;
+
+    if (m_inboundViewTransitionParams && resolveViewTransitionRule())
+        init.viewTransition = ViewTransition::createInbound(*this, std::exchange(m_inboundViewTransitionParams, nullptr));
```
Would be nice if part of these read more like the spec: https://drafts.csswg.org/css-view-transitions-2/#resolve-inbound-cross-document-view-transition

mattwoodrow replied:
Indeed it would, but there's a bunch I haven't implemented there. Will add a FIXME and work on it.

achristensen07 commented on Source/WebCore/dom/PageSwapEvent.h:
```diff
@@ -0,0 +1,55 @@
+    static Ref<PageSwapEvent> create(const AtomString& type, Init&&, IsTrusted = IsTrusted::No);
+    ~PageSwapEvent();
+
+    const RefPtr<NavigationActivation> activation() const { return m_activation; }
```
const RefPtr&

achristensen07 commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -63,7 +63,7 @@ static std::pair<Ref<DOMPromise>, Ref<DeferredPromise>> createPromiseAndWrapper(
 }
 
-ViewTransition::ViewTransition(Document& document, RefPtr<ViewTransitionUpdateCallback>&& updateCallback)
+ViewTransition::ViewTransition(Document& document, std::optional<RefPtr<ViewTransitionUpdateCallback>>&& updateCallback)
```
optional RefPtr is probably an indication that something's not right.  Distinguishing std::nullopt from nullptr is confusing.

achristensen07 commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -805,4 +840,14 @@ bool ViewTransition::documentElementIsCaptured() const
     return renderer->capturedInViewTransition();
 }
 
+std::unique_ptr<ViewTransitionParams> ViewTransition::takeViewTransitionParams()
```
This could return a UniqueRef because it can't return null.

nt1m reviewed (approved):
Non-navigation bits look ok aside from the page update the rendering steps reordering.

I would like @achristensen07 to sign off on the navigation stuff, and it would be good to check if the EWS failures are related.

achristensen07 commented on Source/WebCore/dom/PageRevealEvent.h:
```diff
@@ -0,0 +1,51 @@
+    static Ref<PageRevealEvent> create(const AtomString& type, Init&&, IsTrusted = IsTrusted::No);
+    ~PageRevealEvent();
+
+    const RefPtr<ViewTransition> viewTransition() const { return m_viewTransition; }
```
I would expect this to return a const RefPtr reference instead of an immutable new RefPtr.

mattwoodrow committed:
[css-view-transitions-2] Implement pageswap and pagereveal events.
https://bugs.webkit.org/show_bug.cgi?id=277524
<rdar://133025306>

Reviewed by Tim Nguyen.

Adds the basic IDLs for the pageswap and pagereveal events, plus the C++
implementations.

Adds code to fire these events (gated on the cross document view transitions
preference), and passes the view transition parameter in a partially correct
way.

This is the basic needed for the WPT test infrastructure to work, and as a
result, many new tests pass. There are still lots of FIXME's and bugs left
before the remaining tests pass.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-layer-cascade-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-layer-cascade-external-stylesheet-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-matching-media-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-non-matching-media-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-in-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-multiple-rules-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/at-rule-opt-in-change-with-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/hide-before-reveal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/mismatched-snapshot-containing-block-size-skips-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-microtask-sequence-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-no-view-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-no-view-transition-new-opt-out-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-updatecallbackdone-promise-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pagereveal-with-view-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-long-delay-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-skip-transition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/pageswap-traverse-navigation-no-bfcache.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/skip-outbound-vt-before-reveal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/at-rule-with-types-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/navigation-supersedes-types-same-rule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/navigation-supersedes-types-when-after-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/types-in-pagereveal-and-pageswap-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::reveal):
(WebCore::Document::transferViewTransitionParams):
(WebCore::Document::dispatchPageSwapEvent):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/PageRevealEvent.cpp: Copied from Source/WebCore/page/NavigationActivation.h.
(WebCore::PageRevealEvent::create):
(WebCore::PageRevealEvent::PageRevealEvent):
* Source/WebCore/dom/PageRevealEvent.h: Copied from Source/WebCore/page/NavigationActivation.h.
* Source/WebCore/dom/PageRevealEvent.idl: Added.
* Source/WebCore/dom/PageSwapEvent.cpp: Copied from Source/WebCore/page/NavigationActivation.h.
(WebCore::PageSwapEvent::create):
(WebCore::PageSwapEvent::PageSwapEvent):
* Source/WebCore/dom/PageSwapEvent.h: Copied from Source/WebCore/page/NavigationActivation.h.
* Source/WebCore/dom/PageSwapEvent.idl: Added.
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::ViewTransition):
(WebCore::ViewTransition::create):
(WebCore::ViewTransition::createInbound):
(WebCore::ViewTransition::startInbound):
(WebCore::ViewTransition::callUpdateCallback):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::takeViewTransitionParams):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::OrderedNamedElementsMap::swap):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::navigationCanTriggerCrossDocumentViewTransition):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::clear):
(WebCore::FrameLoader::transitionToCommitted):
* Source/WebCore/page/NavigationActivation.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
(WebCore::operator<<):
* Source/WebCore/page/Page.h:

Canonical link: https://commits.webkit.org/282038@main
